add blog post how to migrate databases

timogoebel opened the pull request:
... this is still WIP as I need to check the commands. Already putting this out here in case anybody wans to check my English. :-)

cc: @lzap, @GregSutcliffe as you asked for this iirc.

timogoebel committed:
add blog post how to migrate databases

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,108 @@
+
+## Motivation
+
+Postgresql and MySQL are both battle-tested open source database management systems. They have specific advantages and disadvantages. The most important reason to prefer one over the other is the available know-how in your organization. If you have a dedicated MySQL team that can provide a database for you, then you should by all means prefer. 
```
The prefer at the end feel like it misses ```that```.

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
````diff
@@ -0,0 +1,108 @@
+If you are using the foreman-tasks plugin, you need to patch the script we are using to migrate the data. If you are not sure, wether you use this plugin, check if you have an entry `Tasks` in Foreman's Monitor menu. If you are still not sure, it is safe to apply the patch in any case.
+
+```sh
+wget -O /tmp/convert-tasks-patch.diff https://gist.github.com/...
````
The URL is missing something. I'd also add a note if this is only needed in 1.15 or similar.

lzap replied:
Yeah, also send patch. I guess it is already pending review somewhere :-)

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,108 @@
+
+## Migrating
+
+To migrate the data from one database to the other, we will use a rake task. You might already be familiar to this tasks from upgrading Foreman to a newer version.
```
s/this tasks/this task/

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,108 @@
+
+## Wrapping up
+
+Postgresql uses sequences in order to store the next free primary key id for a table. These
```
This sentence is incomplete.

ekohl reviewed (commented):
Maybe a note about stopping anything that might start services again (puppet, mon, ..., crons) would be good.

lzap commented:
Great, thanks so much.

With your rich production experience, I would *love* to see some more in-depth summary of performance, some numbers. Does it perform worse, better, first experience? Is there something you are missing now (e.g. replication). These are all great items to have here.

As a redhatter, I only have experience with PostgreSQL Foreman setups...

sean797 commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,108 @@
+- foreman
+---
+
+While Foreman supports both MySQL and Postgresql as a database management system, Katello requires you to use Postgresql. If you want to add the Katello plugin on an existing Foreman setup with MySQL, you need to migrate your database. But there are other reasons why you might want to switch your database management system. This blog post describes what steps are required to migrate your data.
```
s/Katello requires you to use Postgresql/Katello requires Postgresql/ ?

sean797 commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
````diff
@@ -0,0 +1,108 @@
+```yaml
+# The new target database
+development:
+  adapter: sqlite3
````
Should these examples use pgsql and mysql ?

sean797 reviewed (commented):
Nice, 2 small comments :-)

I noticed you haven't mentioned the installer, is that intentional?

timogoebel commented:
@ekohl, @lzap, @sean797: Thanks for your awesome feedback.

I fixed your comments, added the link to the required patch and fixed the examples. Also added some sections about what you suggested.
I forgot, that the user that does the migration needs superuser privileges. Added a section about that.

@lzap: Unfortunately, I don't have any interesting numbers to share. But I tried to describe my first experience. :-(

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,133 @@
+
+First of all, you need to install support for either Postgresql or MySQL to Foreman. One of them will be missing. `yum install -y foreman-postgresql` adds support for Postgresql, `yum install -y foreman-mysql2` does the same for MySQL.
+
+Next, stop all your foreman instances and make sure, they stay stopped for good. On RedHat systems, you need to stop the `httpd` and (if applicable) `foreman-tasks` service. Make sure that neither Puppet or a monitoring system automatically starts these services again.
```
I think this should be ```... neither Puppet nor a ...```

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
````diff
@@ -0,0 +1,133 @@
+If you are using the foreman-tasks plugin, you need to patch the script we are using to migrate the data. If you are not sure, wether you use this plugin, check if you have an entry `Tasks` in Foreman's Monitor menu. If you are still not sure, it is safe to apply the patch in any case.
+
+```sh
+wget -O /tmp/convert-tasks-patch.diff https://gist.githubusercontent.com/timogoebel/cd29e5206678ca0e0d78222af70583ec/raw/bbe998624098c1c3e656fe2582d33b0d9f9285aa/0001-convert-skip-foreman-tasks-tables.patch
````
Is this patch upstream included? Should it be noted that it's only for 1.15?

timogoebel replied:
It's not included upstream. We should provide a way for plugins to specify tables that should be skipped. But that's not trivial to do.
dynflow defines tables that are outside of active record's control. As dynflow is now part of core, I believe we could skip all dynflow tables there. That would be a simple change.

ekohl commented on _posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md:
```diff
@@ -0,0 +1,133 @@
+su - postgres
+vim sequences.sql
+# Paste the script and save the file
+FIXME
```
I think there's something missing here :)

ekohl reviewed (changes requested):
I'd expect something to create the foreman database or at least the user. Should this be documented too?

timogoebel commented:
> I'd expect something to create the foreman database or at least the user. Should this be documented too?

@ekohl: I used puppet for that. But users might use something else, e.g. plain SQL. There is already a ton of information on how this is done on the internet. So I thought it was safe to assume, that this is already done.

Fixed your other comments. Thanks.

ekohl commented:
> @ekohl: I used puppet for that. But users might use something else, e.g. plain SQL. There is already a ton of information on how this is done on the internet. So I thought it was safe to assume, that this is already done.

Since nothing else mentions puppet I'd at least add a note about the installer or ```sudo -u postgres createuser foreman && sudo -u postgres createdb foreman -O foreman -E UTF-8 -l en_US.UTF-8```

lzap reviewed (approved):
Thanks.

dLobatog reviewed (approved):
Since this is already approved & after reading it it looks great, I'm going to go with merging. Thanks @timogoebel & everyone else involved! Time to promote this :)
